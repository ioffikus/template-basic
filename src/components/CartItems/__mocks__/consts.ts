const pendingAPI = 1000;

export { pendingAPI };
