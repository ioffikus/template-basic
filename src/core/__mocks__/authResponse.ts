const authResponse = {
  data: { username: 'user', password: 'password' },
  headers: { 'x-shop-domain': 'template-basic.agg.loc' },
  method: 'POST',
  url: '/login',
};

export default authResponse;
