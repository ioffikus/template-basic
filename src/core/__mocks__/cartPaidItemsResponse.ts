const cartPaidItemsResponse = {
  count: 1,
  next: null as string,
  previous: null as string,
  results: [
    {
      id: 27,
      item: '74212e0f54674aed8964',
      uid: '74212e0f54674aed8964',
      modified: '2017-10-12T09:43:31.442202Z',
      count: 1,
      in_stock: true,
      is_deleted: false,
      schema_fields: {
        name: { type: 'string', label: 'Name', value: 'bla eng bla' },
        short_description: { type: 'text', label: 'Short description', value: 'english bla bla' },
        price: { type: 'decimal', label: 'Price', value: 0 },
        currency: { type: 'string', label: 'Currency', value: '€' },
        picture_url: {
          type: 'string',
          label: 'Picture',
          value:
            'https://d29vc6x3bmmp7p.cloudfront.net/item_pictures/2017-06-05T072530.24449100008df588254089446ab7a52213a70be4c2.jpg',
        },
      },
      aggregion_catalog_item_url: 'agg://collection/58c6ac00a4c8da4900e9024f/bundle',
    },
    {
      id: 108,
      item: 'a9e71d42bab344c9ba1b',
      uid: 'a9e71d42bab344c9ba1b',
      modified: '2017-11-01T10:17:58.183903Z',
      count: 1,
      in_stock: true,
      is_deleted: false,
      schema_fields: {
        name: { type: 'string', value: 'Eng The Last Wish: Introducing (The Witcher)', label: 'Name' },
        short_description: { type: 'text', value: 'Eng description', label: 'Short description' },
        picture_url: {
          type: 'string',
          value:
            'https://d29vc6x3bmmp7p.cloudfront.net/item_pictures/2017-06-05T072535.7689600000119452d41e8941ee85510e6517050b31.jpg',
          label: 'Picture',
        },
        currency: { type: 'string', value: '€', label: 'Currency' },
        price: { type: 'decimal', value: 0, label: 'Price' },
      },
      aggregion_catalog_item_url: 'agg://collection/58d41d54a3fab345003aee61/bundle',
    },
  ],
};

export default cartPaidItemsResponse;
