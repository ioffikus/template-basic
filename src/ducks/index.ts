// instead this, we have router in store via connected-react-router
// import { routerReducer as routing, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';
import auth from './auth';
import cart from './cart';
import cartPaidItems from './cartPaidItems';
import catalogQuery from './catalogQuery';
import catalog from './catalog';
import categories from './categories';
import categoriesMenu from './categoriesMenu';
import filter from './filter';
import locale from './locale';
import meta from './meta';
import page from './page';
import product from './product';
import progress from './progress';
import popularItems from './popularItems';
import recommendItems from './recommendItems';
import routes from './routes';
import ui from './ui';
import user from './user';
import userDelivery from './userDelivery';
import userOrders from './userOrders';

export default combineReducers<Store.IState>({
  ui,
  auth,
  user,
  categories,
  categoriesMenu,
  catalogQuery,
  catalog,
  product,
  progress,
  routes,
  meta,
  filter,
  cart,
  cartPaidItems,
  userDelivery,
  userOrders,
  popularItems,
  recommendItems,
  page,
  locale,
});
