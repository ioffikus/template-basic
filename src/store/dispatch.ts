import { Store } from 'redux';
import _ from 'lodash';
import { showError } from 'src/store/common';

let store: Store<Store.IState>;

export const setupDispatch = (config: { store: Store<Store.IState> }) => {
  store = config.store;
};

const dispatch = (action: any) => {
  if (_.isFunction(action)) {
    return store.dispatch(action).catch((error: any) => {
      if (!_.isNull(error)) {
        showError(error);
      }
      return Promise.reject(error);
    });
  }
  return store.dispatch(action);
};

export default dispatch;
