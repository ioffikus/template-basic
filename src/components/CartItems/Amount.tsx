import React from 'react';
import { connect } from 'react-redux';
import { InputNumber, Spin } from 'antd';
import dispatch from 'src/store/dispatch';
import { increaseQuantity, decreaseQuantity } from 'src/ducks/cart';

interface IOwnProps {
  mayChangeAmountItems: boolean;
  uid: string;
  count: number;
  pendingShowSpinning?: number;
  pendingHideSpinning?: number;
}

interface IConnectedState {}

type IProps = IOwnProps & IConnectedState;

interface IState {
  isSpinning: boolean;
}

const mapStateToProps = (state: Store.IState) => ({});

const WrappedAmount = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Amount extends React.Component<IProps, IState> {
    static defaultProps = {
      pendingShowSpinning: 400,
      pendingHideSpinning: 300,
    };

    spinningTimeoutId: number;

    state: IState = {
      isSpinning: false,
    };

    showSpinning = () => {
      if (!!this.spinningTimeoutId) {
        window.clearTimeout(this.spinningTimeoutId);
        this.spinningTimeoutId = null;
      }
      this.spinningTimeoutId = window.setTimeout(() => {
        this.setState({ isSpinning: true });
      }, this.props.pendingShowSpinning);
    };

    hideSpinning = () => {
      if (!!this.spinningTimeoutId) {
        window.clearTimeout(this.spinningTimeoutId);
        this.spinningTimeoutId = null;
      }
      if (this.state.isSpinning) {
        this.spinningTimeoutId = window.setTimeout(() => {
          this.setState({ isSpinning: false });
        }, this.props.pendingHideSpinning);
      }
    };

    handleChangeAmount = (value: number) => {
      if (process.env.BROWSER) {
        const { uid, count } = this.props;
        this.showSpinning();
        if (count < value) {
          dispatch(
            increaseQuantity(uid, () => {
              this.hideSpinning();
            }),
          );
        } else {
          dispatch(
            decreaseQuantity(uid, () => {
              this.hideSpinning();
            }),
          );
        }
      }
    };

    componentWillUnmount() {
      if (process.env.BROWSER) {
        if (!!this.spinningTimeoutId) {
          window.clearTimeout(this.spinningTimeoutId);
        }
      }
    }

    render() {
      const { mayChangeAmountItems, count } = this.props;
      return (
        <div className="root">
          <style jsx>{`
            .root {
              display: inline-block;
              vertical-align: top;
            }
            .root :global(.ant-input-number) {
              border: 0;
              display: flex;
              justify-content: center;
              background: transparent;
              color: #062f49;
            }
            .root :global(.ant-input-number:hover) {
              border-color: transparent;
            }
            .root :global(.ant-input-number-focused) {
              box-shadow: none;
              border-color: transparent;
            }
            .root :global(.ant-input-number-handler-wrap) {
              border: 0;
              opacity: 1;
              width: 100%;
              display: flex;
              flex-direction: row-reverse;
              background: transparent;
            }
            .root :global(.ant-input-number-input-wrap) {
              position: relative;
              z-index: 3;
              width: 32px;
              height: 32px;
            }
            .root :global(.ant-input-number-handler) {
              width: 50%;
              height: 100%;
              border: 0;
              top: 0;
              bottom: 0;
              background: transparent;
            }
            .root :global(.ant-input-number-handler:hover) {
              height: 100% !important;
            }
            .root :global(.ant-input-number-handler-down-inner) {
              right: auto;
              left: 4px;
            }
            .root :global(.ant-input-number-input) {
              background-color: #e5ecf4;
              border-radius: 8px;
              color: inherit;
              text-align: right;
              font-weight: bold;
            }
          `}</style>
          {mayChangeAmountItems ? (
            <Spin spinning={this.state.isSpinning}>
              <InputNumber min={1} value={count} onChange={this.handleChangeAmount} />
            </Spin>
          ) : (
            count
          )}
        </div>
      );
    }
  },
);

export default WrappedAmount;
