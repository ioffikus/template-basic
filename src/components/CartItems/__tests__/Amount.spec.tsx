import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { InputNumber, Spin } from 'antd';
import configureMockStore from 'redux-mock-store';
import _ from 'lodash';
import { setupDispatch } from 'src/store/dispatch';
import { increaseQuantity, decreaseQuantity } from 'src/ducks/cart';
import { pendingAPI } from '../__mocks__/consts';
import Amount from '../Amount';

let increaseQuantityCb: () => void;
let decreaseQuantityCb: () => void;
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
jest.mock('src/ducks/cart', () => {
  const { pendingAPI } = require('../__mocks__/consts');
  return {
    increaseQuantity: jest.fn((_, cb) => {
      increaseQuantityCb = cb;
      return () =>
        new Promise(resolve => {
          setTimeout(() => {
            cb();
            resolve();
          }, pendingAPI);
        });
    }),
    decreaseQuantity: jest.fn((_, cb) => {
      decreaseQuantityCb = cb;
      return () =>
        new Promise(resolve => {
          setTimeout(() => {
            cb();
            resolve();
          }, pendingAPI);
        });
    }),
  };
});

describe(null, () => {
  it('render() with mayChangeAmountItems as true', () => {
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = 123;
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Amount mayChangeAmountItems={mayChangeAmountItems} uid={uid} count={count} />, {
      context: { store },
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('render() with mayChangeAmountItems as false', () => {
    const mayChangeAmountItems = false;
    const uid = 'dummy';
    const count = 123;
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Amount mayChangeAmountItems={mayChangeAmountItems} uid={uid} count={count} />, {
      context: { store },
    });
    expect(wrapper.dive()).toMatchSnapshot();
  });
  it('InputNumber.onChange() with increaseQuantity()', () => {
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = 1;
    const value = 2;
    expect(count).toBeLessThan(value);
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Amount mayChangeAmountItems={mayChangeAmountItems} uid={uid} count={count} />, {
      context: { store },
    });
    const dummyWrapper = wrapper.dive();
    const inputNumberInstance = dummyWrapper
      .find(InputNumber)
      .dive()
      .instance() as InputNumber;
    inputNumberInstance.props.onChange(value);
    expect(increaseQuantity).toBeCalledWith(uid, increaseQuantityCb);
  });
  it('InputNumber.onChange() with decreaseQuantity()', () => {
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = 2;
    const value = 2;
    expect(count).toBeGreaterThanOrEqual(value);
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(<Amount mayChangeAmountItems={mayChangeAmountItems} uid={uid} count={count} />, {
      context: { store },
    });
    const dummyWrapper = wrapper.dive();
    const inputNumberInstance = dummyWrapper
      .find(InputNumber)
      .dive()
      .instance() as InputNumber;
    inputNumberInstance.props.onChange(value);
    expect(decreaseQuantity).toBeCalledWith(uid, decreaseQuantityCb);
  });
  it('InputNumber.onChange() with pendingAPI < pendingShowSpinning (dont show Spin)', () => {
    jest.useFakeTimers();
    const pendingShowSpinning = pendingAPI + 200;
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = _.random(1, 2);
    const value = _.random(1, 2);
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(
      <Amount
        pendingShowSpinning={pendingShowSpinning}
        mayChangeAmountItems={mayChangeAmountItems}
        uid={uid}
        count={count}
      />,
      {
        context: { store },
      },
    );
    const dummyWrapper = wrapper.dive();
    const inputNumberInstance = dummyWrapper
      .find(InputNumber)
      .dive()
      .instance() as InputNumber;
    inputNumberInstance.props.onChange(value);
    jest.runTimersToTime(pendingShowSpinning + 100); // + 100 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperBefore = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperBefore.props.spinning).toBeFalsy();
    jest.runTimersToTime(pendingAPI + 100); // + 100 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperAfter = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperAfter.props.spinning).toBeFalsy();
    jest.clearAllTimers();
    jest.useRealTimers();
  });
  it('InputNumber.onChange() without pendingAPI < pendingShowSpinning (show Spin)', () => {
    jest.useFakeTimers();
    const pendingShowSpinning = pendingAPI - 200;
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = _.random(1, 2);
    const value = _.random(1, 2);
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(
      <Amount
        pendingShowSpinning={pendingShowSpinning}
        mayChangeAmountItems={mayChangeAmountItems}
        uid={uid}
        count={count}
      />,
      {
        context: { store },
      },
    );
    const dummyWrapper = wrapper.dive();
    const inputNumberInstance = dummyWrapper
      .find(InputNumber)
      .dive()
      .instance() as InputNumber;
    inputNumberInstance.props.onChange(value);
    jest.runTimersToTime(pendingShowSpinning + 100); // + 100 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperBefore = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperBefore.props.spinning).toBeTruthy();
    jest.runTimersToTime(pendingAPI + 100); // + 100 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperAfter = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperAfter.props.spinning).toBeFalsy();
    jest.clearAllTimers();
    jest.useRealTimers();
  });
  it('InputNumber.onChange() with double click before show Spin', () => {
    jest.useFakeTimers();
    const timerOverlapOne = -200;
    const timerOverlapTwo = -200;
    const timerOverlapThree = 100;
    const pendingShowSpinning = pendingAPI / 3;
    expect(pendingShowSpinning * 3 + timerOverlapTwo + timerOverlapTwo + timerOverlapThree).toBeLessThan(pendingAPI);
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = _.random(1, 2);
    const value = _.random(1, 2);
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(
      <Amount
        pendingShowSpinning={pendingShowSpinning}
        mayChangeAmountItems={mayChangeAmountItems}
        uid={uid}
        count={count}
      />,
      {
        context: { store },
      },
    );
    const dummyWrapper = wrapper.dive();
    const inputNumberInstance = dummyWrapper
      .find(InputNumber)
      .dive()
      .instance() as InputNumber;
    inputNumberInstance.props.onChange(value);
    jest.runTimersToTime(pendingShowSpinning + timerOverlapOne); // - 200 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperBeforeOne = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperBeforeOne.props.spinning).toBeFalsy();
    inputNumberInstance.props.onChange(value);
    jest.runTimersToTime(pendingShowSpinning + timerOverlapTwo); // - 200 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperBeforeTwo = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperBeforeTwo.props.spinning).toBeFalsy();
    inputNumberInstance.props.onChange(value);
    jest.runTimersToTime(pendingShowSpinning + timerOverlapThree); // + 100 for timer overlap
    dummyWrapper.update();
    const spinInstanceWrapperAfter = dummyWrapper
      .find(Spin)
      .dive()
      .instance() as Spin;
    expect(spinInstanceWrapperAfter.props.spinning).toBeTruthy();
    jest.clearAllTimers();
    jest.useRealTimers();
  });
  it('componentWillUnmount() with two calls of clearTimeout()', () => {
    jest.useFakeTimers();
    const pendingShowSpinning = pendingAPI - 200;
    const mayChangeAmountItems = true;
    const uid = 'dummy';
    const count = _.random(1, 2);
    const value = _.random(1, 2);
    const store = mockStore();
    setupDispatch({ store });
    const wrapper = shallow(
      <Amount
        pendingShowSpinning={pendingShowSpinning}
        mayChangeAmountItems={mayChangeAmountItems}
        uid={uid}
        count={count}
      />,
      {
        context: { store },
      },
    );
    const dummyWrapper = wrapper.dive();
    const inputNumberInstance = dummyWrapper
      .find(InputNumber)
      .dive()
      .instance() as InputNumber;
    inputNumberInstance.props.onChange(value);
    jest.runTimersToTime(pendingAPI + 100); // + 100 for timer overlap
    dummyWrapper.unmount();
    const clearTimeoutMock = _.get(window.clearTimeout, 'mock', null);
    expect(clearTimeoutMock.calls.length).toBe(2);
    jest.clearAllTimers();
    jest.useRealTimers();
  });
});
