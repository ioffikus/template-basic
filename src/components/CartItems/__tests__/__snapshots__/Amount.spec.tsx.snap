// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() with mayChangeAmountItems as false 1`] = `
<div
  className="jsx-1853710974 root"
>
  <JSXStyle
    css=".root.jsx-1853710974{display:inline-block;vertical-align:top;}.root.jsx-1853710974 .ant-input-number{border:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-webkit-justify-content:center;-ms-flex-pack:center;justify-content:center;background:transparent;color:#062f49;}.root.jsx-1853710974 .ant-input-number:hover{border-color:transparent;}.root.jsx-1853710974 .ant-input-number-focused{box-shadow:none;border-color:transparent;}.root.jsx-1853710974 .ant-input-number-handler-wrap{border:0;opacity:1;width:100%;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row-reverse;-ms-flex-direction:row-reverse;flex-direction:row-reverse;background:transparent;}.root.jsx-1853710974 .ant-input-number-input-wrap{position:relative;z-index:3;width:32px;height:32px;}.root.jsx-1853710974 .ant-input-number-handler{width:50%;height:100%;border:0;top:0;bottom:0;background:transparent;}.root.jsx-1853710974 .ant-input-number-handler:hover{height:100% !important;}.root.jsx-1853710974 .ant-input-number-handler-down-inner{right:auto;left:4px;}.root.jsx-1853710974 .ant-input-number-input{background-color:#e5ecf4;border-radius:8px;color:inherit;text-align:right;font-weight:bold;}"
    styleId="1853710974"
  />
  123
</div>
`;

exports[`render() with mayChangeAmountItems as true 1`] = `
<div
  className="jsx-1853710974 root"
>
  <JSXStyle
    css=".root.jsx-1853710974{display:inline-block;vertical-align:top;}.root.jsx-1853710974 .ant-input-number{border:0;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-pack:center;-webkit-justify-content:center;-ms-flex-pack:center;justify-content:center;background:transparent;color:#062f49;}.root.jsx-1853710974 .ant-input-number:hover{border-color:transparent;}.root.jsx-1853710974 .ant-input-number-focused{box-shadow:none;border-color:transparent;}.root.jsx-1853710974 .ant-input-number-handler-wrap{border:0;opacity:1;width:100%;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-flex-direction:row-reverse;-ms-flex-direction:row-reverse;flex-direction:row-reverse;background:transparent;}.root.jsx-1853710974 .ant-input-number-input-wrap{position:relative;z-index:3;width:32px;height:32px;}.root.jsx-1853710974 .ant-input-number-handler{width:50%;height:100%;border:0;top:0;bottom:0;background:transparent;}.root.jsx-1853710974 .ant-input-number-handler:hover{height:100% !important;}.root.jsx-1853710974 .ant-input-number-handler-down-inner{right:auto;left:4px;}.root.jsx-1853710974 .ant-input-number-input{background-color:#e5ecf4;border-radius:8px;color:inherit;text-align:right;font-weight:bold;}"
    styleId="1853710974"
  />
  <Spin
    prefixCls="ant-spin"
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <InputNumber
      min={1}
      onChange={[Function]}
      prefixCls="ant-input-number"
      step={1}
      value={123}
    />
  </Spin>
</div>
`;
